### PR TITLE
feat(circleci): add support for CircleCI Personal Access Token and Project Access Token. PRP #1671

### LINKED ISSUE #1671: PRP: Secret extractor for CircleCI
* **Secret name**: `Circle CI Personal Access Token`

* **Risk in exposing the secret**:
An exposed CCIPAT token would allow an attacker to impersonate the
associated CircleCI user and perform a wide range of privileged operations.
An attacker could interact with and potentially modify:

* **Jobs** (trigger, list, rerun, or manipulate job behavior)
* **Projects** (enumerate, query configuration, or alter project-level data)
* **Keys** (list or access SSH and API keys)
* **Artifacts** (retrieve or tamper with build artifacts)
* **Environment Variables**

* List Environment Variables
* Add Environment Variables
* Get a Single Environment Variable
* Delete Environment Variables
This can lead to pipeline compromise, code execution, theft of sensitive
secrets, supply-chain manipulation, and unauthorized operations across
CI/CD workflows.

* **Validation method, if any**:

* Perform a lightweight, non-destructive validation request using the token:

```bash
curl https://circleci.com/api/v1.1/me -H "Circle-Token: <circle-token>"
```

A valid token returns identifying information such as:

```json
{
"basic_email_prefs": "smart",
"login": "pbiggar"
}
```
* If the API returns user metadata, the token is confirmed to be active and
tied to a real CircleCI account.
* Submissions that include an automated validation Enricher plugin for this
endpoint are rewarded more highly.

* **Resources**:

https://circleci.com/docs/api/v2/

### PR DESCRIPTION
**Pull request by learningat2002**
Resolves #1671

### CODE REVIEW

File: veles/secrets/circleci/detector.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+const (
+	patMaxLen     = 100 // CCIPAT_ (7) + identifier + _ + 40 hex chars
+	projectMaxLen = 100 // CCIPRJ_ (7) + identifier + _ + 40 hex chars
+)
+
+// circleCIPATRe matches CircleCI Personal Access Tokens in the format:
+// CCIPAT_[A-Za-z0-9]+_[0-9a-f]{40}
+var circleCIPATRe = regexp.MustCompile(`CCIPAT_[A-Za-z0-9]+_[0-9a-f]{40}`)
```
Comment (erikvarga) [inline veles/secrets/circleci/detector.go]: Does the identifier part have a specific length or is it always varying? If we know the length ranges we can better ensure that patMaxLen is set properly

File: veles/secrets/circleci/detector_test.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+	testPAT     = "CCIPAT_GHFzqc7fRZ2GviZQ7hbdeb_9f54ac82eef4bb69a8fece88199a7414f32d8b36"
+	testProject = "CCIPRJ_Nw1xCXXyTW8uvdkHKLNUqK_4ad9cadd8b2b29d02a49ed03720fac5644f66c92"
+)
+
+func TestPersonalAccessTokenDetectorAcceptance(t *testing.T) {
+	velestest.AcceptDetector(
+		t,
+		circleci.NewPersonalAccessTokenDetector(),
+		testPAT,
+		circleci.PersonalAccessToken{Token: testPAT},
```
Comment (erikvarga) [inline veles/secrets/circleci/detector_test.go:38]: Also configure WithBackToBack and WithPad('A') to test that the regexps don't rely on word boundaries. Example:

https://github.com/google/osv-scalibr/blob/93e36a8c27c6a04fe491934bf26169c6d42027b4/veles/secrets/onepasswordkeys/detector_test.go#L58

File: veles/secrets/circleci/detector_test.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name  string
+		input string
+	}{
+		{
+			name:  "Invalid prefix",
+			input: `CCIPAT_invalid`,
```
Comment (erikvarga) [inline veles/secrets/circleci/detector_test.go]: the hex length is also invalid here so let's change that to be a valid length

File: veles/secrets/circleci/detector_test.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name  string
+		input string
+	}{
+		{
+			name:  "Invalid prefix",
+			input: `CCIPRJ_invalid`,
```
Comment (erikvarga) [inline veles/secrets/circleci/detector_test.go]: the hex length is also invalid here so let's change that to be a valid length

File: veles/secrets/circleci/detector_test.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name  string
+		input string
+		want  []veles.Secret
+	}{
+		{
+			name:  "PAT in config",
```
Comment (erikvarga) [inline veles/secrets/circleci/detector_test.go]: Use underscores instead of spaces in test names: https://github.com/google/osv-scalibr/blob/main/docs/style_guide.md#use-easy-to-find-subtest-descriptions

File: veles/secrets/circleci/validator.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+func statusFromProjectTokenResponseBody(body io.Reader) (veles.ValidationStatus, error) {
+	var resp projectTokenErrorResponse
+	if err := json.NewDecoder(body).Decode(&resp); err != nil {
+		// Decoding failed -> ambiguous response, treat as failed to validate.
+		return veles.ValidationFailed, fmt.Errorf("unable to parse response: %w", err)
+	}
+
+	// CircleCI returns 404 with "Not Found" when the token is valid but the project doesn't exist.
```
Comment (erikvarga) [inline veles/secrets/circleci/validator.go]: Can we use just the HTTP status code to tell this error case apart from others where the token is not valid? If so then let's just check the status code by filling out ValidResponseCodes/InvalidResponseCodes since the response body is more likely to change with API updates than the status code.

File: veles/secrets/circleci/validator.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+	// Other error messages (like "Invalid token provided.") indicate the token is invalid.
+	return veles.ValidationInvalid, nil
+}
+
+// NewProjectTokenValidator creates a new CircleCI Project Token Validator.
+//
+// Validation approach:
+// CircleCI Project tokens are validated by attempting to access a non-existent project.
+// We use HTTP Basic Auth with the token as username and empty password.
```
Comment (erikvarga) [inline veles/secrets/circleci/validator.go]: If the password is left empty is the authentication still successful using just the token?

File: veles/secrets/circleci/validator_test.go [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+	"testing"
+
+	"github.com/google/osv-scalibr/veles"
+	"github.com/google/osv-scalibr/veles/secrets/circleci"
+)
+
+func TestPersonalAccessTokenValidator_Valid(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Header.Get("Circle-Token") == testPAT {
+			w.WriteHeader(http.StatusOK)
```
Comment (erikvarga) [inline veles/secrets/circleci/validator_test.go]: The main difference between this test and TestPersonalAccessTokenValidator_Invalid status seems to be the returned HTTP status and the expected validation. Let's merge them into one test and use table-driven testing to set the returned HTTP status and the expected validation result.

File: veles/secrets/circleci/validator_test.go [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+	status, err := validator.Validate(context.Background(), secret)
+	if err != nil {
+		t.Fatalf("Validate() error = %v, want nil", err)
+	}
+	if status != veles.ValidationInvalid {
+		t.Errorf("Validate() status = %v, want %v", status, veles.ValidationInvalid)
+	}
+}
+
+func TestProjectTokenValidator_Valid_200(t *testing.T) {
```
Comment (erikvarga) [inline veles/secrets/circleci/validator_test.go]: We can probably merge these 3 tests into a single table-driven test in a similar way, by keeping the server logic from here and replacing the returned status + expected validation results with the test parameters.

File: veles/secrets/circleci/detector_test.go (hunk 5 above)
Comment (learningat2002) [reply veles/secrets/circleci/detector_test.go]: Sorry !

File: veles/secrets/circleci/validator.go (hunk 7 above)
Comment (learningat2002) [reply veles/secrets/circleci/validator.go]: yes, but changed to use header with token for more clarity.

File: veles/secrets/circleci/detector_test.go (hunk 3 above)
Comment (learningat2002) [reply veles/secrets/circleci/detector_test.go]: done

**Review by erikvarga — APPROVED** [review]
